scripts/stage_layer.py

```diff
@@ -0,0 +1,94 @@
+from argparse import ArgumentParser
+import logging
+from pathlib import Path
+import subprocess
+import sys
+from zipfile import (
+    ZipFile,
+    ZipInfo,
+    ZIP_DEFLATED,
+)
+
+import boto3
+
+from azul import config
+from azul.logging import configure_script_logging
+
+log = logging.getLogger(__name__)
+
+
+class LayerBuilder:
+    layer_dir = Path(config.project_root) / 'lambdas' / 'layer'
+    out_dir = layer_dir / '.chalice' / 'terraform'
+
+    def __init__(self):
+        self.client = boto3.client('s3')
+
+    def update_layer_if_necessary(self, bucket, key):
+        log.info('Updating layer in bucket: %s with key %s', bucket, key)
+        if self.update_required(bucket, key):
+            log.info('Updating layer.zip')
+            self.update_layer(bucket, key)
+        else:
+            log.info('No update necessary')
+
+    def update_required(self, bucket, key) -> bool:
+        log.info('Checking if layer.zip needs to be updated')
+        try:
+            # Since the object key is content-addressable, just checking for
+            # the file's presence is sufficient
+            self.client.head_object(Bucket=bucket, Key=key)
+        except self.client.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == '404':
+                return True
+            else:
+                raise
+        else:
+            return False
+
+    def update_layer(self, bucket, key):
+        input_zip = self.out_dir / 'deployment.zip'
+        layer_zip = self.out_dir / 'layer.zip'
+        log.info('Tainting old lambda layer')
+        command = ['terraform', 'taint', 'aws_lambda_layer_version.dependencies']
+        subprocess.run(command, cwd=Path(config.project_root) / 'terraform')
+        self.build_layer_zip(layer_zip, input_zip)
+        log.info('Uploading layer zip to S3')
+        self.client.upload_file(str(layer_zip), bucket, key)
```

Comment:
Since we removed the stage name from the key, it might make sense to [tag](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-tagging.html) the object with the deployment, just so we can easily find the layer used by a deployment, when looking at the objects in the bucket directly. @natanlao is adding generic tagging so the tag name should be consistent with what he chose.

Reply:
I'd rather not do this for a couple reasons:
- it's already easy to find which layer is connected to which lambda from the lambda console
- since layers are now shared between deployments this creates a race on updating the tag. If the tag key contained the deployment name there would be no race, but I think this would be ugly.

Reply:
Sounds good.